File: strings/determine-if-two-strings-are-close/determine-if-two-strings-are-close.go
package determineiftwostringsareclose

import "sort"

func closeStrings(word1 string, word2 string) bool {
	if len(word1) != len(word2) {
		return false
	}

	chars1 := make([]int, 26)
	chars2 := make([]int, 26)

	for i := 0; i < len(word1); i++ {
		chars1[word1[i]-'a']++
		chars2[word2[i]-'a']++
	}

	for i := 0; i < len(chars1); i++ {
		if (chars1[i] == 0) != (chars2[i] == 0) {
			return false
		}
	}

	sort.Ints(chars1)
	sort.Ints(chars2)

	for i := 0; i < len(chars1); i++ {
		if chars1[i] != chars2[i] {
			return false
		}
	}

	return true
}

File: strings/determine-if-two-strings-are-close/determine-if-two-strings-are-close.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {boolean}
 */
function closeStrings(word1, word2) {
  if (word1.length !== word2.length) {
    return false;
  }

  const chars1 = new Array(26).fill(0);
  const chars2 = new Array(26).fill(0);

  for (let i = 0; i < word1.length; i++) {
    chars1[word1[i].charCodeAt(0) - 97]++;
    chars2[word2[i].charCodeAt(0) - 97]++;
  }

  for (let i = 0; i < chars1.length; i++) {
    if ((chars1[i] === 0) !== (chars2[i] === 0)) {
      return false;
    }
  }

  chars1.sort((a, b) => a - b);
  chars2.sort((a, b) => a - b);

  for (let i = 0; i < chars1.length; i++) {
    if (chars1[i] !== chars2[i]) {
      return false;
    }
  }
  return true;
}
